infrastructure/crypto/src/hash.rs

```diff
@@ -0,0 +1,28 @@
+// Copyright 2018 The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE
+
+/// A trait for structs that need to produce a canonical hash
+pub trait Hashable {
```

Comment:
Isnt there already a hashable trait in crypto?

Reply:
There is but its only for the built in hashing the standard library uses for HashMap and HashSet so it doesn't support choosing your Hasher.

Reply:
I propose dropping this trait. 

It's a super-thin wrapper around `Digest` so I suggest just using that wherever we use `Hashable` (digest and hash are basically synonyms) or DRYing up `crypto::Challenge`  if you find yourself adding functionality that will likely repeat what's already there.

Reply:
So this trait does something that is not catered for by `Digest` which is that allows one to define how a complex structure is hashed up (in what order and how are the various members turned into byte streams and concatenated together etc.). The standard library does have a Hashable trait but we would still need to create wrappers for all the fancy `Digest` functions that have been implemented in `RustCrypto` in any case so I think this is a neater solution.

My feeling is that stuff in `Challenge` is essentially exactly what the Digest trait offers so I do agree that `Challenge` can be DRYed up. I will create an Issue to examine that refactor.